File: wordle-java/src/Wordle.java
// A simple program to demonstrate
// Tic-Tac-Toe Game.
import java.util.*;

public class Wordle {

    static String userWordToGuess;
    static Map<Character, Set<Integer>> lettersWithPlacements;
    static boolean startNewRound = true;

    static Scanner in = new Scanner(System.in);

    static final int totalTurns = 7;
    static final int lettersInWord = 5;


    public static void main(String[] args)
    {
        while (startNewRound) {
            newRound();
        }
        in.close();
    }

    public static void newRound() {
        getWordAndCreateLetterMap();
        System.out.println("Guess a " + lettersInWord + "-letter word. Enter q to quit the current round at any time");
        playGame();
        System.out.println();
        System.out.println();
    }

    public static void playGame() {
        String userInput;

        int numCorrectlyPlacedLetters;
        int [] resultsOfUserInput = new int [lettersInWord]; // 0 = not in word, 1 = in word but wrong placements, 2 = correct placement

        boolean stillPlayingCurrentRound = true;
        boolean guessedCorrectly = false;
        boolean quitGame = false;
        int numTurnsLeft = totalTurns;

        while (stillPlayingCurrentRound) {
            System.out.println("You have " + numTurnsLeft + " guesses");
            userInput = in.nextLine();
            if (userInput.equals("q")) {
                stillPlayingCurrentRound = false;
                quitGame = true;
                continue;
            }
            if (userInput.length() != lettersInWord) {
                System.out.println("Please enter a " + lettersInWord + "-letter word");
                continue;
            }
            Arrays.fill(resultsOfUserInput, 0);
            numCorrectlyPlacedLetters = 0;
            for (int i = 0; i < lettersInWord; i ++) {
                char letter = userInput.charAt(i);
                if (lettersWithPlacements.containsKey(letter)) {
                    if (lettersWithPlacements.get(letter).contains(i)) {
                        resultsOfUserInput[i] = 2;
                        numCorrectlyPlacedLetters ++;
                    } else {
                        resultsOfUserInput[i] = 1;
                    }
                } else {
                    resultsOfUserInput[i] = 0;
                }
            }

            for (int number : resultsOfUserInput) {
                System.out.print(number + " ");
            }
            System.out.println();

            if (numCorrectlyPlacedLetters == lettersInWord) {
                stillPlayingCurrentRound = false;
                guessedCorrectly = true;
            }

            numTurnsLeft --;
            if (numTurnsLeft == 0) {
                stillPlayingCurrentRound = false;
                guessedCorrectly = false;
            }
        }
        if (quitGame) {
            System.out.println("You have entered q to quit the current round.");
        }
        else if (guessedCorrectly) {
            System.out.println("Congrats! The word was " + userWordToGuess);
        } else {
            System.out.println("You did not guess the word correctly. The word was " + userWordToGuess);
        }
        System.out.println("Press a to re-start a new round, and Press b to quit the game entirely");
        userInput = in.nextLine();
        if (userInput.equals("b")) {
            startNewRound = false;
        }
    }

    public static void getWordAndCreateLetterMap() {
        // Get new word
        // Get the instance of APIDataManager
        WordsFromApi wordsFromApi = WordsFromApi.getInstance();
        String word = wordsFromApi.getRandomWordFromList();
        userWordToGuess = word;

        //setup hash map
        lettersWithPlacements = new HashMap<>();
        for (int i = 0; i < word.length(); i++) {
            char letter = word.charAt(i);
            lettersWithPlacements.computeIfAbsent(letter, k -> new HashSet<>()).add(i);
        }
    }
}

File: wordle-kotlin/src/main/kotlin/Wordle.kt
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import java.net.URL

@Serializable
data class WordFromApiData(
    @SerialName("word") val word: String,
    @SerialName("score") val score: Int
)

class DataMuseApiCaller {
    companion object {
        private var wordList: MutableList<String> = mutableListOf()

        fun getWordList(): List<String> {
            return wordList.toList()
        }

        fun getRandomWordFromList(): String {
            return wordList.random();
        }

        suspend fun setWordList() {
            try {
                val apiUrl = "https://api.datamuse.com/words?sp=?????"

                var jsonResponse = withContext(Dispatchers.IO) {
                    URL(apiUrl).readText()
                }

                val wordDataList: List<WordFromApiData> = Json.decodeFromString(jsonResponse)

                wordList.addAll(wordDataList.map { it.word })
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}

class GameStatusValues {
    companion object {
        var startNewRound: Boolean = true
        var wordForUserToGuess: String = ""
        var lettersWithPlacements: MutableMap<Char, MutableSet<Int>> = HashMap()
        val totalTurns: Int = 7
        val lettersInWord: Int = 5
    }
}

fun getWordAndCreateLetterMap() {
    val word = DataMuseApiCaller.getRandomWordFromList()
    GameStatusValues.lettersWithPlacements = HashMap()
    for (i in word.indices) {
        val letter = word[i]
        GameStatusValues.lettersWithPlacements.computeIfAbsent(letter) { HashSet() }.add(i)
    }
    GameStatusValues.wordForUserToGuess = word
}

fun playGame() {
//    println(GameStatusValues.wordForUserToGuess)
//    println(GameStatusValues.lettersWithPlacements)


    var numCorrectlyPlacedLetters: Int
    var resultsOfUserInput: MutableList<Int>
    var stillPlayingCurrentRound = true
    var guessedCorrectly = false
    var quitGame = false
    var numTurnsLeft: Int = GameStatusValues.totalTurns

    var userInput: String?

    while (stillPlayingCurrentRound) {
        println("You have $numTurnsLeft left")
        userInput = readLine()
        if (userInput == "q") {
            stillPlayingCurrentRound = false
            quitGame = true
            continue
        }
        if (userInput == null || userInput.length != GameStatusValues.lettersInWord) {
            println("Please enter a ${GameStatusValues.lettersInWord}-letter word")
            continue
        }
        numCorrectlyPlacedLetters = 0
        resultsOfUserInput = mutableListOf()
        for (i in 0 until GameStatusValues.lettersInWord) {
            val letter = userInput[i]
            if (GameStatusValues.lettersWithPlacements.containsKey(letter)) {
                if (GameStatusValues.lettersWithPlacements[letter]?.contains(i) == true) {
                    resultsOfUserInput.add(2)
                    numCorrectlyPlacedLetters ++
                } else {
                    resultsOfUserInput.add(1)
                }
            } else {
                resultsOfUserInput.add(0)
            }
        }

        for (number in resultsOfUserInput) {
            print("$number ")
        }
        println()

        if (numCorrectlyPlacedLetters == GameStatusValues.lettersInWord) {
            guessedCorrectly = true
            stillPlayingCurrentRound = false
        }

        numTurnsLeft --
        if (numTurnsLeft == 0) {
            guessedCorrectly = false
            stillPlayingCurrentRound = false
        }
    }

    if (quitGame) {
        println("You have entered q to quit the current round.")
    } else if (guessedCorrectly) {
        println("Congrats! The word was ${GameStatusValues.wordForUserToGuess}")
    } else {
        println("You did not guess the word correctly. The word was ${GameStatusValues.wordForUserToGuess}")
    }
    println("Press a to re-start a new round, and Press b to quit the game entirely")
    userInput = readLine()
    if (userInput == "b") {
        GameStatusValues.startNewRound = false
    }
}

fun newRound() {
    getWordAndCreateLetterMap()
    println("Guess a ${GameStatusValues.lettersInWord}-letter word. Enter q to quit the current round at any time")
    playGame()
    println()
    println()
}

suspend fun main() {
    DataMuseApiCaller.setWordList()
    while (GameStatusValues.startNewRound) {
        newRound()
    }
}
